storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3MultiPartOutputStream.java

```diff
@@ -167,11 +178,14 @@ private void abortUpload() {
         closed = true;
     }
 
-    private void flushBuffer(final int offset,
+    private void flushBuffer(final ByteBuffer buffer,
                              final int actualPartSize) {
-        final ByteArrayInputStream in = new ByteArrayInputStream(partBuffer.array(), offset, actualPartSize);
-        uploadPart(in, actualPartSize);
-        partBuffer.clear();
+        try (final InputStream in = new ByteBufferInputStream(buffer)) {
+            processedBytes += actualPartSize;
+            uploadPart(in, actualPartSize);
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
Also, provide an inputstream directly from buffer stream instead of creating a new byte array.